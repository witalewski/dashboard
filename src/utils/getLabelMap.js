export const getLabelMap = source => ({
  label: 'label',
  ...Object.keys(source)
    .filter(key => key.match(/-label/))
    .reduce((acc, key) => ({ ...acc, [key.split('-')[0]]: source[key] }), {}),
});
