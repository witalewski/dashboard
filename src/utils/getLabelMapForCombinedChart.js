export const getLabelMapForCombinedChart = source => ({
  label: 'label',
  ...Object.keys(source)
    .filter(key => key.match(/-bar|-label/))
    .reduce((acc, key) => ({ ...acc, [key.split('-')[0]]: source[key] }), {}),
});
