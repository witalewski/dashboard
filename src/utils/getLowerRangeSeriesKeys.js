export const getLowerRangeSeriesKeys = (labelMap, processedData) => {
  const sortedSeries = Object.values(labelMap)
    .filter(e => e !== 'label')
    .map(key => ({
      key: key,
      avgValue:
        processedData.reduce((acc, item) => acc + item[key], 0) /
        processedData.length,
    }))
    .sort((a, b) => a.avgValue - b.avgValue);
  const halfOfSortedSeriesLength = Math.floor(sortedSeries.length / 2);
  const lowerRangeSeries = sortedSeries.slice(0, halfOfSortedSeriesLength);
  if (sortedSeries.length % 2 !== 0) {
    let middleItem = sortedSeries[halfOfSortedSeriesLength];
    if (
      Math.abs(
        middleItem.avgValue - sortedSeries[halfOfSortedSeriesLength - 1]
      ) <
      Math.abs(middleItem.avgValue - sortedSeries[halfOfSortedSeriesLength + 1])
    ) {
      lowerRangeSeries.push(middleItem);
    }
  }
  return lowerRangeSeries.map(item => item.key);
};
