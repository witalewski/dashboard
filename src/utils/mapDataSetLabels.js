export const mapDataSetLabels = (labelMap, data) =>
  data.map(item =>
    Object.keys(item).reduce(
      (acc, key) => ({ ...acc, [labelMap[key]]: item[key] }),
      {}
    )
  );
