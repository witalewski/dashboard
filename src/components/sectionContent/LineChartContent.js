import React from 'react';
import {
  LineChart,
  Line,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
} from 'recharts';
import { chartColors } from '../../global/styleConstants';
import { mapDataSetLabels } from '../../utils/mapDataSetLabels';

export const LineChartContent = ({ data, labelMap }) => {
  const processedData = mapDataSetLabels(labelMap, data);
  return (
    <LineChart
      width={600}
      height={300}
      data={processedData}
      margin={{ top: 5, right: 30, left: 20, bottom: 5 }}
    >
      <XAxis dataKey={Object.keys(data[0])[0]} />
      <YAxis />
      <CartesianGrid strokeDasharray="3 3" />
      <Tooltip />
      <Legend />
      {Object.keys(processedData[0])
        .slice(1)
        .map((key, i) => (
          <Line
            key={key}
            type="monotone"
            dataKey={key}
            stroke={chartColors[i % chartColors.length]}
          />
        ))}
    </LineChart>
  );
};
