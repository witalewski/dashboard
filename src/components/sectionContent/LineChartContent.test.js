import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import mockChartData from '../../fixtures/mockChartData.json';
import { LineChartContent } from './LineChartContent';
import { getLabelMap } from '../../utils/getLabelMap';

describe('LineChartContent', () => {
  it('matches snapshot with mock data', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <LineChartContent
        title="Revenue This Month vs Revenue Previous Month"
        className="revenue-section-wrapper"
        data={mockChartData.revenueMonthToMonth.data}
        labelMap={getLabelMap(mockChartData.revenueMonthToMonth)}
      />
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
