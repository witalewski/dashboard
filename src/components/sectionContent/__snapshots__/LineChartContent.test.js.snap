// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChartContent matches snapshot with mock data 1`] = `
<LineChart
  barCategoryGap="10%"
  barGap={4}
  data={
    Array [
      Object {
        "Days of Month": 1,
        "Previous Month": 2156.75,
        "This Month": 34997.48,
      },
      Object {
        "Days of Month": 2,
        "Previous Month": 4688.6,
        "This Month": 34997.12,
      },
      Object {
        "Days of Month": 3,
        "Previous Month": 12045.43,
        "This Month": 49449.76,
      },
      Object {
        "Days of Month": 4,
        "Previous Month": 33516.68,
        "This Month": 62774.3,
      },
    ]
  }
  height={300}
  layout="horizontal"
  margin={
    Object {
      "bottom": 5,
      "left": 20,
      "right": 30,
      "top": 5,
    }
  }
  reverseStackOrder={false}
  stackOffset="none"
  width={800}
>
  <XAxis
    allowDataOverflow={false}
    allowDecimals={true}
    allowDuplicatedCategory={true}
    dataKey="x"
    domain={
      Array [
        0,
        "auto",
      ]
    }
    height={30}
    hide={false}
    mirror={false}
    orientation="bottom"
    padding={
      Object {
        "left": 0,
        "right": 0,
      }
    }
    reversed={false}
    scale="auto"
    tickCount={5}
    type="category"
    width={0}
    xAxisId={0}
  />
  <YAxis
    allowDataOverflow={false}
    allowDecimals={true}
    allowDuplicatedCategory={true}
    domain={
      Array [
        0,
        "auto",
      ]
    }
    height={0}
    hide={false}
    mirror={false}
    orientation="left"
    padding={
      Object {
        "bottom": 0,
        "top": 0,
      }
    }
    reversed={false}
    scale="auto"
    tickCount={5}
    type="number"
    width={60}
    yAxisId={0}
  />
  <CartesianGrid
    fill="none"
    horizontal={true}
    horizontalFill={Array []}
    horizontalPoints={Array []}
    stroke="#ccc"
    strokeDasharray="3 3"
    vertical={true}
    verticalFill={Array []}
    verticalPoints={Array []}
  />
  <Tooltip
    active={false}
    animationDuration={400}
    animationEasing="ease"
    contentStyle={Object {}}
    coordinate={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    cursor={true}
    cursorStyle={Object {}}
    filterNull={true}
    isAnimationActive={true}
    itemSorter={[Function]}
    itemStyle={Object {}}
    labelStyle={Object {}}
    offset={10}
    separator=" : "
    useTranslate3d={false}
    viewBox={
      Object {
        "x1": 0,
        "x2": 0,
        "y1": 0,
        "y2": 0,
      }
    }
    wrapperStyle={Object {}}
  />
  <Legend
    align="center"
    iconSize={14}
    layout="horizontal"
    verticalAlign="bottom"
  />
  <Line
    activeDot={true}
    animateNewValues={true}
    animationBegin={0}
    animationDuration={1500}
    animationEasing="ease"
    connectNulls={false}
    dataKey="Previous Month"
    dot={true}
    fill="#fff"
    hide={false}
    isAnimationActive={true}
    legendType="line"
    onAnimationEnd={[Function]}
    onAnimationStart={[Function]}
    points={Array []}
    stroke="#5db990"
    strokeWidth={1}
    type="monotone"
    xAxisId={0}
    yAxisId={0}
  />
  <Line
    activeDot={true}
    animateNewValues={true}
    animationBegin={0}
    animationDuration={1500}
    animationEasing="ease"
    connectNulls={false}
    dataKey="This Month"
    dot={true}
    fill="#fff"
    hide={false}
    isAnimationActive={true}
    legendType="line"
    onAnimationEnd={[Function]}
    onAnimationStart={[Function]}
    points={Array []}
    stroke="#864ba5"
    strokeWidth={1}
    type="monotone"
    xAxisId={0}
    yAxisId={0}
  />
</LineChart>
`;
